Edited: OpenACC port of the Split-Explicit implementation.

Pranay-Reddy-Kommera opened the pull request:
Edited: This PR includes OpenACC changes to Split-Explicit (SE) implementation. Below is the summary of the changes:
- Ported all the routines invoked in the SE implementations based on namelist configuration used.
- Additional OpenACC kernel optimizations are implemented.

NOTE: This PR also require the changes from [MPAS-framework#16](https://github.com/EarthWorksOrg/mpas-framework/pull/16) to enable appropriate data transfers.

Commit:
OpenACC port of both the Split-explicit and Split-implicit implementations.

dazlich commented:
Ok, first I wanted to check that this will build on the cpu. src/mode_forward/mpas_ocn_time_integration_si.F had issues. I made the following changes:

   use cublas

          becomes

#ifdef USE_CUBLAS
   use cublas
#endif

   type(cublasHandle) :: h

          becomes

#ifdef USE_CUBLAS
   type(cublasHandle) :: h
#endif

      istat = cublasCreate(h)

          becomes

#ifdef USE_CUBLAS
      istat = cublasCreate(h)
#endif

Now it builds.

dazlich commented:
Next step, I have built the code for CPU with the nvhpc compiler - does it run?

split_explicit runs and gives the same solution as before. Split_implicit is blowing up, right off the bat.

Pranay-Reddy-Kommera commented:
> Member
I haven't tested a CPU build. I will test the CPU build and keep you updated.

dazlich commented:
@Pranay-Reddy-Kommera - here are the modifications to enable the cpu-build:

dazlich@derecho6:~/earthworks/openacc/components/mpas-ocean/src/mode_forward> git diff mpas_ocn_time_integration_si.F

    integer :: istat
 !$acc declare create(prec_ivmat,SIcst_allreduce_local2,SIcst_allreduce_global2,SIcst_allreduce_local3,SIcst_allreduce_global3,&
 !$acc                SIcst_allreduce_local9,SIcst_allreduce_global9,globalReprodSum2fld1,globalReprodSum2fld2,globalReprodSum3fld1,&
@@ -2638,7 +2642,9 @@ module ocn_time_integration_si
       'MPAS was not compiled with LAPACK/BLAS. LAPACK required for SI' &
       , MPAS_LOG_CRIT)
 #endif
+#ifdef USE_CUBLAS
       istat = cublasCreate(h)
+#endif
 
       !*** Set mask for using velocity correction
       if (config_vel_correction) then
@@ -3425,6 +3431,12 @@ module ocn_time_integration_si
                           SIvec_in(1:nPrecVec) , 1, 0.0_RKIND,  &
                           SIvec_out(1:nPrecVec), 1)
 !$acc end host_data
+#else
+#ifdef USE_LAPACK
+      call DSPMV('U', nPrecVec, 1.0_RKIND, prec_ivmat, &
+                 SIvec_in(1:nPrecVec) , 1, 0.0_RKIND,  &
+                 SIvec_out(1:nPrecVec), 1)
+#endif
 #endif
       elseif ( trim(config_btr_si_preconditioner) == &
                                         'block_jacobi' ) then
@@ -3443,6 +3455,12 @@ module ocn_time_integration_si
                           SIvec_in(1:nPrecVec) , 1, 0.0_RKIND,  &
                           SIvec_out(1:nPrecVec), 1)
 !$acc end host_data
+#else
+#ifdef USE_LAPACK
+      call DSPMV('U', nPrecVec, 1.0_RKIND, prec_ivmat, &
+                 SIvec_in(1:nPrecVec) , 1, 0.0_RKIND,  &
+                 SIvec_out(1:nPrecVec), 1)
+#endif
 #endif
       elseif ( trim(config_btr_si_preconditioner) == &
                                               'jacobi' ) then

dazlich committed:
updates to enable cpu build and run

Places where cublas is used need preprocessor directives so the code will continue to build on cpu. In a couple places cpu code had been commented and needed to be restored to run on cpu. These places are part of #ifdef USE_CUBLAS blocks.

dazlich commented:
@Pranay-Reddy-Kommera - the above comment is not quite complete. But I've done a commit that you can look at and try.

Now I am moving on to the gpu build and run.

Commit:
Reverted back the changes to Split-Implicit Implementation.

Pranay-Reddy-Kommera commented:
@dazlich @sjsprecious  I have modified the PR to only include changes to the Split-Explicit implementation. Could you please review the changes.

dazlich committed:
Modify for gpu compilation

This commit explicitly puts the compiler flags for
gpu compilation in here. This is kludgy and needs
generalization.

dazlich commented:
@Pranay-Reddy-Kommera Ok, I've had a chance to try this out. Works fine on cpu. For gpu it was not compiling with the gpu flags. I've committed a somewhat ugly buildlib for mpas-ocean that does compile with the flags. It needs generalization, but at this point, since we are working with standalone ocean edit in buildlib to adjust gpu compiler flags. Unfortunately, the gpu code crashes in the initialization phase.

Pranay-Reddy-Kommera commented:
@dazlich Could you please let me know what is the error you see when it is crashing in the initialization phase?
Just to confirm, please let me know if you have also included https://github.com/EarthWorksOrg/mpas-framework/pull/16 while running the code. We need both the PRs for MPAS-Ocean to run.

dazlich commented:
@Pranay-Reddy-Kommera - yes I included the PR from maps-framework.

cesm.log has the following for each PE:

deg0066.hsn.de.hpc.ucar.edu 36: malloc: call to cuMemAllocManaged returned error 1 (CUDA_ERROR_INVALID_VALUE): Invalid value
deg0066.hsn.de.hpc.ucar.edu 36: 0: ALLOCATE: 0 bytes requested

dazlich commented:
@Pranay-Reddy-Kommera -
my case directory is /glade/derecho/scratch/dazlich/mpaso_gpu_120km.openacc.se
my source directory is ~dazlich/earthworks/openacc2
my run script is ~dazlich/gpu_o120.csh

dazlich committed:
Removed the managed memory compiler flag

dazlich commented:
@Pranay-Reddy-Kommera - I've rerun without the managed memory. Now I get the model crashing during the initialization and producing diagnostic output with NaNs in the prognostic fields.
